File: src-cpp/leetcode/container-with-most-water.cpp
#include <bits/stdc++.h>
using namespace std;

int maxArea(vector<int> &height)
{
    int n = height.size();
    int i = 0, j = n - 1;
    int maxRectArea = INT_MIN;

    while (i < j)
    {
        maxRectArea = max(maxRectArea, min(height[i], height[j]) * (j - i));

        if (height[i] < height[j])
            i++;
        else
            j--;
    }

    return maxRectArea;
}

// brute force
int maxArea(vector<int> &height)
{
    int n = height.size();
    int maxRectArea = INT_MIN;

    for (int i = 0; i < n; i++)
    {
        for (int j = i + 1; j < n; j++)
        {
            maxRectArea = max(maxRectArea, min(height[i], height[j]) * (j - i));
        }
    }

    return maxRectArea;
}

File: src-ts/leetcode/container-with-most-water.ts
function maxArea(height: number[]): number {
  let maxA = -Infinity;
  const n = height.length;

  let i = 0;
  let j = n - 1;

  while (i < j) {
    const h = Math.min(height[i], height[j]);

    maxA = Math.max(maxA, h * (j - i));

    while (height[i] <= h && i < j) {
      i += 1;
    }

    while (height[j] <= h && i < j) {
      j -= 1;
    }
  }

  return maxA;
}
